### PR TITLE
WIP Move event loop into JavaScript.

### PR DESCRIPTION
**Pull request by ry**

### CODE REVIEW

**Comment by ry** [comment]
Before:
```
gpu ~/src/deno> ./tools/throughput_benchmark.py out/before/deno 5000
listening on 127.0.0.1:4544
head -c 5242880000 /dev/zero | nc 127.0.0.1 4544
34.2299828529 seconds
```
After:
```
gpu ~/src/deno> ./tools/throughput_benchmark.py out/after/deno 5000
listening on 127.0.0.1:4544
head -c 5242880000 /dev/zero | nc 127.0.0.1 4544
18.5639469624 seconds
```


Running `strace -c ./out/release/deno tests/fetch_deps.ts --allow-net`

Before this patch:
```
% time     seconds  usecs/call     calls    errors syscall
------ ----------- ----------- --------- --------- ----------------
 99.90    0.003860         161        24           futex
  0.10    0.000004           0       634           mprotect
  0.00    0.000000           0        22           read
  0.00    0.000000           0         3           write
  0.00    0.000000           0        16           open
  0.00    0.000000           0        19           close
  0.00    0.000000           0         2           stat
  0.00    0.000000           0        15           fstat
  0.00    0.000000           0        60           mmap
  0.00    0.000000           0        63           munmap
  0.00    0.000000           0         2           brk
  0.00    0.000000           0         5           rt_sigaction
  0.00    0.000000           0         1           rt_sigprocmask
  0.00    0.000000           0         5         1 ioctl
  0.00    0.000000           0         9         9 access
  0.00    0.000000           0        34        17 madvise
  0.00    0.000000           0         7           clone
  0.00    0.000000           0         1           execve
  0.00    0.000000           0         1           getcwd
  0.00    0.000000           0         2         2 mkdir
  0.00    0.000000           0         1         1 readlink
  0.00    0.000000           0         2           getrlimit
  0.00    0.000000           0         3           sigaltstack
  0.00    0.000000           0         1           arch_prctl
  0.00    0.000000           0         4           sched_getaffinity
  0.00    0.000000           0         1           set_tid_address
  0.00    0.000000           0         1           epoll_ctl
  0.00    0.000000           0         1           set_robust_list
  0.00    0.000000           0         1           epoll_create1
  0.00    0.000000           0         1           pipe2
  0.00    0.000000           0         4           getrandom
------ ----------- ----------- --------- --------- ----------------
100.00    0.003864                   945        30 total
```

After:
```
% time     seconds  usecs/call     calls    errors syscall
------ ----------- ----------- --------- --------- ----------------
  0.00    0.000000           0        22           read
  0.00    0.000000           0         3           write
  0.00    0.000000           0        16           open
  0.00    0.000000           0        19           close
  0.00    0.000000           0         2           stat
  0.00    0.000000           0        15           fstat
  0.00    0.000000           0        59           mmap
  0.00    0.000000           0       634           mprotect
  0.00    0.000000           0        59           munmap
  0.00    0.000000           0         2           brk
  0.00    0.000000           0         5           rt_sigaction
  0.00    0.000000           0         1           rt_sigprocmask
  0.00    0.000000           0         5         1 ioctl
  0.00    0.000000           0         9         9 access
  0.00    0.000000           0        34        17 madvise
  0.00    0.000000           0         7           clone
  0.00    0.000000           0         1           execve
  0.00    0.000000           0         1           getcwd
  0.00    0.000000           0         2         2 mkdir
  0.00    0.000000           0         1         1 readlink
  0.00    0.000000           0         2           getrlimit
  0.00    0.000000           0         3           sigaltstack
  0.00    0.000000           0         1           arch_prctl
  0.00    0.000000           0        23           futex
  0.00    0.000000           0         4           sched_getaffinity
  0.00    0.000000           0         1           set_tid_address
  0.00    0.000000           0         1           epoll_ctl
  0.00    0.000000           0         1           set_robust_list
  0.00    0.000000           0         1           epoll_create1
  0.00    0.000000           0         1           pipe2
  0.00    0.000000           0         2           getrandom
------ ----------- ----------- --------- --------- ----------------
100.00    0.000000                   937        30 total
```

**Comment by kevinkassimo** [comment]
Probably we need a benchmark dedicated for futex. Looks pretty bad

**Comment by ztplz** [comment]
Here is same issue. [https://groups.google.com/forum/#!topic/comp.unix.programmer/5RAMxJJnRcE](url)

**Comment by ry** [comment]
@kevinkassimo futex is most likely from [the std::sync::mpsc we use in the main thread](https://github.com/denoland/deno/blob/32f07971284f05993cf702aa589859e7900406b4/src/isolate.rs#L223-L227). It looks bad, and I'm not sure it's the best design, but I doubt this is the source of slowness. (I'd rather do epoll the main thread - but it will require [making a custom Tokio runtime](https://tokio.rs/docs/going-deeper/building-runtime/).)
Likely the source of slowness is from flatbuffer serialization on the JS side. It allocates a new ArrayBuffer for each message, and this patch makes each poll serialize a message. I'm going to try to validate this hypothesis with the profiler before trying to fix it. More updates soon...

**Comment by ry** [comment]
FYI Here's how I am profiling:
```
# Make sure we're only building release.
> export DENO_BUILD_MODE=release
# Build deno and V8's d8.
> ./tools/build.py d8 deno
# Start the program we want to benchmark with --prof
> ./out/release/deno tests/http_bench.ts --allow-net --prof &
# Exercise it.
> third_party/wrk/linux/wrk http://localhost:4500/
# Kill server.
> kill `pgrep deno`
# When supplying --prof, V8 will write a file in the current directory that looks 
# like this isolate-0x7fad98242400-v8.log
# To examine this file
> D8_PATH=out/release/ ./third_party/v8/tools/linux-tick-processor isolate-0x7fad98242400-v8.log
```

This produces a profile: https://gist.github.com/ry/65c9e620225f3d1829cfe39e88cb8467

**Comment by ry** [comment]
Counter-intuitively this patch makes the program spend more time in C++. Look at the summary:

```
# BEFORE
 [Summary]:
   ticks  total  nonlib   name
   1780   20.3%   21.1%  JavaScript
   6606   75.3%   78.2%  C++
    373    4.3%    4.4%  GC
    325    3.7%          Shared libraries
     62    0.7%          Unaccounted

# AFTER
 [Summary]:
   ticks  total  nonlib   name
    985    9.6%    9.8%  JavaScript
   8981   87.1%   89.7%  C++
    391    3.8%    3.9%  GC
    296    2.9%          Shared libraries
     46    0.4%          Unaccounted
```

File: js/dispatch.ts [hunk 1]
```diff
@@ -5,35 +5,86 @@ import * as msg from "gen/msg_generated";
...
+      }
+    }
+    eventLoopLog();
+    const ui8 = poll(delay);
+    if (ui8 != null) {
+      handleAsyncMsgFromRust(ui8);
+      libdeno.runMicrotasks();
+    }
+  }
+  return promiseErrorExaminer();
```
Comment (kevinkassimo) [inline js/dispatch.ts]: Just curious, does this mean that we will only handle promise uncaught rejects when deno is about to exit?

Comment (kevinkassimo) [reply js/dispatch.ts]: Hmm. I think with this impl deno would not promptly die on uncaught errors after running to the end.

Which one was the original failing test?
(quite busy in the coming weeks but might as well take a look)

Comment (ry) [reply js/dispatch.ts]: tests/async_error.ts

File: js/dispatch.ts [hunk 2]
```diff
@@ -5,35 +5,86 @@ import * as msg from "gen/msg_generated";
...
+  util.log(`TICK delay ${delay} nTasks ${nTasks}`);
+}
+
+function idle(): boolean {
+  delay = fireTimers();
+  return delay < 0 && nTasks === 0;
+}
+
+export function eventLoop(): boolean {
+  for (;;) {
```
Comment (kitsonk) [inline js/dispatch.ts]: I always liked `while (true)` better for infinite loops, feels more intentional to me, but...

**Comment by ry** [comment]
Closing this one. It resulted in profiling and ongoing perf fixes. The ultimate goal is to move the loop to JS - but we're not ready to do it yet.

**Pull request closed without merging**